File: heroic-component/src/main/java/com/spotify/heroic/grammar/ListExpression.java
/*
 * Copyright (c) 2015 Spotify AB.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.spotify.heroic.grammar;

import com.fasterxml.jackson.annotation.JsonTypeName;
import com.google.common.base.Joiner;
import lombok.Data;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

@Data
@JsonTypeName("list")
public final class ListExpression implements Expression {
    private final Context context;
    private final List<Expression> list;

    @Override
    public ListExpression eval(final Scope scope) {
        final List<Expression> list = Expression.evalList(this.list, scope);
        return new ListExpression(context, list);
    }

    @Override
    public <R> R visit(final Visitor<R> visitor) {
        return visitor.visitList(this);
    }

    @Override
    public ListExpression add(Expression other) {
        final ListExpression o = other.cast(ListExpression.class);
        final ArrayList<Expression> list = new ArrayList<Expression>();
        list.addAll(this.list);
        list.addAll(o.list);
        return new ListExpression(context.join(other.getContext()), list);
    }

    @Override
    public String toRepr() {
        final Joiner arg = Joiner.on(", ");
        final Iterator<String> args = list.stream().map(Expression::toRepr).iterator();
        return "[" + arg.join(args) + "]";
    }
}

File: heroic-component/src/main/java/com/spotify/heroic/grammar/ListExpression.kt
/*
 * Copyright (c) 2019 Spotify AB.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"): you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.spotify.heroic.grammar

data class ListExpression(@JvmField val context: Context, val list: List<Expression>): Expression {
    override fun getContext() = context

    override fun eval(scope: Expression.Scope): Expression {
        val list = Expression.evalList(list, scope)
        return ListExpression(context, list)
    }

    override fun <R : Any?> visit(visitor: Expression.Visitor<R>): R {
        return visitor.visitList(this)
    }

    override fun add(other: Expression): ListExpression {
        val o: ListExpression = other.cast(ListExpression::class.java)
        val list = listOf(*this.list.toTypedArray(), *o.list.toTypedArray())

        return ListExpression(context.join(other.context), list)
    }

    override fun toRepr() = list.joinToString(", ", transform = Expression::toRepr)
}
